src/common/utils.js

```diff
@@ -140,6 +140,80 @@ function promiseAny(promises: Promise<any>[]): Promise<any> {
   })
 }
 
+/**
+ * Waits for the promises to resolve,
+ * until the callback function is happy with the results,
+ * or there are no more promises.
+ */
+async function promiseNy<T, Out>(
+  promises: Promise<T>[],
+  cleanerFunc: (result: T) => Out | void,
+  n?: number = promises.length
+): Promise<Out> {
+  const results: T[] = []
+  function checkResults(results: T[]): Out | void {
+    if (results.length >= n) {
+      const sortedResults = results
+        .map(result => JSON.stringify(cleanerFunc(result)))
+        .sort()
+      for (let i = 0; i < sortedResults.length; i++) {
+        for (let j = 0; j < sortedResults.length; j++) {
+          if (i === j) continue
+          if (sortedResults[i] !== sortedResults[j]) break
+          if (j === n - 1) return JSON.parse(sortedResults[i])
+        }
+      }
+    }
+    return undefined
+  }
+
+  return new Promise((resolve, reject) => {
+    let failed = 0
+    let done = false
+    for (const promise of promises) {
+      promise.then(
+        result => {
+          results.push(result)
+          const checked = checkResults(results)
+          if (!done && checked != null) {
+            done = true
+            resolve(checked)
+          }
+        },
+        error => {
+          if (++failed + results.length >= promises.length) {
+            reject(error)
+          }
+        }
+      )
+    }
+  })
+}
```

Comment:
In summary, change this promiseNy utility function to accept a compareFn argument that will be used to check the equality of two results. Use a resultCountMap object (`new Map`), to keep a running count on the number of "equivalent" result objects. Once a count for a particular result has exceeded the `n` parameter, then exit the algorithm and resolve the promise.

Rought, example code:

```
promise.then(result => {
  for (const [existingResult, count] of resultCountMap.entries()) {
    if (compare(existingResult, result)) {
      resultCountMap.set(existingResult, count + 1)
    }

    if (resultCountMap.get(existingResult) >= n) {
      return resolve(result)
    }
  }
})
```